### ISSUE TITLE
[TS] Make type definitions for Locale more specific

### ISSUE THREAD

**Issue by vkbansal** [date-fns/date-fns#612]
Right now, the type definitions for `Locale` are very generic and almost everything is `Function`. https://github.com/date-fns/date-fns/blob/master/typings.d.ts#L44

Can it be updated like
```ts
weekdays: () => Array<string>
months: () => Array<string>
```

I would like to help with this.

**Comment by leshakoss** [comment]
@vkbansal what do you mean? These values are functions right now

**Comment by vkbansal** [comment]
Yes these are functions but they do not define argument types and return type of the function.

For Example, `weekdays` and `months` are functions that return an array of strings, but right now they are defined as `Function`s which is not very helpful when using TypeScript.

```ts
import { localize as en } from 'date-fns/locale/en-US';
const monthNames = en.months();
```
As you can see below, upon inspection it shows `months()` as `Function`:
<img width="493" alt="screen shot 2017-11-16 at 2 15 40 pm" src="https://user-images.githubusercontent.com/5930351/32881957-d616b896-cad8-11e7-97e9-1b6133606a92.png">

and `monthNames` is showing `any`:
<img width="408" alt="screen shot 2017-11-16 at 2 16 01 pm" src="https://user-images.githubusercontent.com/5930351/32881960-d7682b8a-cad8-11e7-9704-82be50c4b4db.png">

Ideally `monthNames` should show `Array<string>` and `months()` as `() => Array<string>`

You can refer https://www.typescriptlang.org/docs/handbook/functions.html

**Comment by leshakoss** [comment]
Oh alright, I get it. Yes, it would be useful :-) The typings are generated from this file: https://github.com/date-fns/date-fns/blob/master/docs/Locale.js by this script: https://github.com/date-fns/date-fns/blob/master/scripts/build/typings.js

Beware: it certainly needs some refactoring 🙃 Also I am not really sure that [JSDoc](http://usejsdoc.org) supports verbose function signatures

**Comment by leshakoss** [comment]
Oh, typing are actually generated by two scripts. There is how you do it:
```sh
./scripts/build/docs.js && ./scripts/build/typings.js
```
It also could be quite slow

**Comment by vkbansal** [comment]
Cool I'll take a look.

**Comment by vkbansal** [comment]
> Also I am not really sure that JSDoc supports verbose function signatures

I think this can be used http://usejsdoc.org/tags-callback.html

**Comment by kangax** [comment]
@vkbansal I also just noticed how loose types are in date-fns. Have you ever made any progress on making them stricter?

**Comment by vkbansal** [comment]
I had looked into it, but was unsuccessful.

**Comment by KonradLinkowski** [comment]
Still not fixed
Generated type:
```
type Locale = {
  code?: string
  formatDistance?: (...args: Array<any>) => any
  formatRelative?: (...args: Array<any>) => any
  localize?: {
    ordinalNumber: (...args: Array<any>) => any
    era: (...args: Array<any>) => any
    quarter: (...args: Array<any>) => any
    month: (...args: Array<any>) => any
    day: (...args: Array<any>) => any
    dayPeriod: (...args: Array<any>) => any
  }
  formatLong?: {
    date: (...args: Array<any>) => any
    time: (...args: Array<any>) => any
    dateTime: (...args: Array<any>) => any
  }
  match?: {
    ordinalNumber: (...args: Array<any>) => any
    era: (...args: Array<any>) => any
    quarter: (...args: Array<any>) => any
    month: (...args: Array<any>) => any
    day: (...args: Array<any>) => any
    dayPeriod: (...args: Array<any>) => any
  }
  options?: {
    weekStartsOn?: 0 | 1 | 2 | 3 | 4 | 5 | 6
    firstWeekContainsDate?: 1 | 2 | 3 | 4 | 5 | 6 | 7
  }
}
```
Actual type:
https://github.com/date-fns/date-fns/blob/master/src/locale/types.ts#L17

**Comment by josean-dev** [comment]
Also encountering this issue, really weird 

> Still not fixed Generated type:
> 
> ```
> type Locale = {
>   code?: string
>   formatDistance?: (...args: Array<any>) => any
>   formatRelative?: (...args: Array<any>) => any
>   localize?: {
>     ordinalNumber: (...args: Array<any>) => any
>     era: (...args: Array<any>) => any
>     quarter: (...args: Array<any>) => any
>     month: (...args: Array<any>) => any
>     day: (...args: Array<any>) => any
>     dayPeriod: (...args: Array<any>) => any
>   }
>   formatLong?: {
>     date: (...args: Array<any>) => any
>     time: (...args: Array<any>) => any
>     dateTime: (...args: Array<any>) => any
>   }
>   match?: {
>     ordinalNumber: (...args: Array<any>) => any
>     era: (...args: Array<any>) => any
>     quarter: (...args: Array<any>) => any
>     month: (...args: Array<any>) => any
>     day: (...args: Array<any>) => any
>     dayPeriod: (...args: Array<any>) => any
>   }
>   options?: {
>     weekStartsOn?: 0 | 1 | 2 | 3 | 4 | 5 | 6
>     firstWeekContainsDate?: 1 | 2 | 3 | 4 | 5 | 6 | 7
>   }
> }
> ```
> 
> Actual type: https://github.com/date-fns/date-fns/blob/master/src/locale/types.ts#L17

Encountering this same issue. Why are the generated types different?

**Comment by KonradLinkowski** [comment]
@josean-dev types are generated using jsdoc, typescript definitions are ignored while building for some reason. I tried to look into that, but the process is complicated
